write CRLF instead of escaped form for Newline

mike-kfed opened the pull request:
I wrote a tool that changes the font-size of existing RTF files. When using the `to_rtf()` method to write the new RTF I had to make a workaround to actually write out newlines, because the current code writes `\r\n` to the files instead of the actual CR LF ascii codes.

mike-kfed committed:
write CRLF instead of escaped form for Newline

compenguy commented:
I see my mistake now.

The specification says this:
```
A carriage return (character value 13) or linefeed (character value 10) will be treated as a \par control if the character is preceded by a backslash. You must include the backslash; otherwise, RTF ignores the control word. (You may also want to insert a carriage-return/linefeed pair without backslashes at least every 255 characters for better text transmission over communication lines.)
```

What I did was record a literal backslash-literal r-literal backslash-literal n, when I should have recorded a literal backslash-escaped r-literal backslash-escaped n, or in other words `\\\r\\\n`, according to the specification, or possibly better `\\\u{13}\\\u{10}`.

But the specification definitely says that a rendered newline should be encoded with a backslash preceding the ascii codes.

mike-kfed commented:
well I am ignorant to the specs ;) I just opened the source rtf and my output rtf in a text editor. The source was shown with newlines, the output was one single line. looking at the files in rtf viewer showed different text, namely additional `\r\n` so I changed how the `Token::Newline` is written and all is fine.
I think there are two cases of Newline, one is the DOS style CRLF sequence from text files, which is recommended at least every 255 chars per line, and another one that is interpreted like `\par` that's the one with the preceding backslashes.

I played around with this, indeed within curly braces here hex escaped for illustration:
`{\pard ... example.\x5c\x0d\x5c\x0a}` indeed behaves like a `\par` but my rtf files actually use `\par` and the newlines to make it easier to read the file source in an editor.
I propose therefore to add another enum variant `Token::ParNewline` that parses in this weird backslash before CR LF and writes out `\par` to not be stupid?

mike-kfed commented:
side note: the `\x5c\x0d\x5c\x0a` sequence also behaves like `\par` outside the curly braces. It is very weird to read in a text editor and I personally see no advantage in writing that back out to files again - I'd write out `\par` instead. So correctly interpret when deserialising, get rid of it when serialising.

side-side note: only `\x5c\x0d` also works for the macOS preview app, making this even weirder. And the full sequence behaves like two `\par` (i.e. two lines of vertical space before the next text)

compenguy commented:
Ok, it's been a long time since I really looked at this, and it's slowly coming back to me now.  We want to be permissive in parsing, which we are.

If you look in `src/raw.rs`:

```
// Oddly enough, the copy of the RTF spec we have has at least one carriage return without its
// matching line feed, so it looks like we need to be more permissive about newlines than the spec
// says.                                
pub fn newline_raw(input: &[u8]) -> IResult<&[u8], &[u8]> {
    alt((crlf, tag("\n"), tag("\r")))(input)
}                 
```

We accept `CR LF`, `CR`, and `LF` (but not `LF CR`, or rather, that would be treated as two separate newlines, for better or for worse).  

Unfortunately, the macOS behavior of rendering CRLF as two line breaks is just Apple being Apple.  macOS, by historical convention, uses CR as the line break character.  Unix, by historical convention, uses LF.  Windows uses CRLF.  The RTF preview is apparently choosing to interpret the CRLF as a macOS line break followed by a Unix line break.  It's spec-breaking, and trying to adapt to making it render correctly on macOS would probably break it elsewhere.

I think in order to make sure RTF documents round-trip correctly (what we write out matches what we read in, or at least relatively closely), we may need to do two things.  First, make the Newline token store the newline characters that were read in.  The second, is to make the to_rtf function write out the stored newline characters as we read them.

```
pub fn read_newline(input: &[u8]) -> IResult<&[u8], Token> {
    map(newline_raw, Token::Newline)(input)                 
}  

#[derive(PartialEq, Eq)]                       
pub enum Token {                      
    ControlSymbol(char),              
    ControlWord {                          
        name: String,                                           
        arg: Option<i32>,                                    
    },                                                              
    ControlBin(Vec<u8>),                                            
    /// Text is not str because it can be in any of various encodings -
    /// it's up to the processor to identify any encoding information in
    /// the stream, and do any encoding conversion desired        
    Text(Vec<u8>),                      
    StartGroup,                                                        
    EndGroup,                                                           
    Newline(String),                                                                
}                                        

    pub fn to_rtf(&self) -> Vec<u8> {
        match self {
            Token::ControlSymbol(c) => format!("\\{c}").as_bytes().to_vec(),
            Token::ControlWord { name, arg } => match arg {
                Some(num) => format!("\\{name}{num}").as_bytes().to_vec(),
                None => format!("\\{name}").as_bytes().to_vec(),
            },
            Token::ControlBin(data) => {
                let mut rtf: Vec<u8> = Vec::with_capacity(12 + data.len());
                rtf.extend_from_slice(format!("\\bin{} ", data.len()).as_bytes());
                rtf.extend_from_slice(data);
                rtf
            }
            Token::Text(data) => data.to_vec(),
            Token::StartGroup => b"{".to_vec(),
            Token::EndGroup => b"}".to_vec(),
            Token::Newline(s) => s.as_bytes().to_vec(),
        }
    }
```

That's not all the changes that would be required - there's a couple other spots where Token::Newline gets used that you'd need update to account for the String value now associated with it.  But I think that would clean things up nicely.  If you read an input document, and it contained just CR newlines, that's what you'll get in the output.  If it contained `\CRLF', then that would actually be stored as a different token type, not actually the Newline one, and that would also be written back out correctly, I think.  And in the macOS case, the output document should render the same as the input one.

mike-kfed commented:
Yeah I also figured that could be a historic macOS CR thing, and that I shouldn't use Apple viewer as a reference ;). Good idea with storing the original newline found in the enum variant. I'll implement that and make it work. thanks for the quick responses!

mike-kfed committed:
store the newline chars found, add test for ControlSymbol \CR

mike-kfed commented:
after re-reading the RTF specs you posted above, I think that Apple is doing the right thing, it says backslash prefix either CR or LF behaves like \par. Meaning have \CR followed by \LF is supposed to be interpreted as 2 times \par.

Also doing tests I realised that your library handles this backslash case with the ControlSymbol variant. So I just added a test for it to the existing string.

I added the parsing and dumping of newline style anyways, for my RTF files parse then write out all tokens now makes a perfect copy (unless my program is instructed to change the font size of course).

compenguy commented:
Published as v0.2.1

mike-kfed commented:
> Published as v0.2.1

amazing, thanks!
